core/src/main/scala/cats/MonadError.scala

```diff
@@ -13,6 +13,23 @@ trait MonadError[F[_], E] extends ApplicativeError[F, E] with Monad[F] {
   def ensure[A](fa: F[A])(error: => E)(predicate: A => Boolean): F[A] =
     flatMap(fa)(a => if (predicate(a)) pure(a) else raiseError(error))
 
+  /**
+   * Sequences the specified finalizer ensuring evaluation regardless of
+   * whether or not the target `F[A]` raises an error.
+   *
+   * If `raiseError` is analogous to `throw` and `handleErrorWith` is analogous to
+   * `catch`, then `guarantee` is analogous to `finally`.  JVM exception semantics
+   * are mirrored with respect to error raised within the finalizer (i.e. errors
+   * raised within the finalizer will shade any errors raised by the primary action).
+   *
+   * @see [[raiseError]]
+   * @see [[handleErrorWith]]
+   */
+  def guarantee[A](fa: F[A], finalizer: F[Unit]): F[A] = {
```

Comment:
@djspiewak If `guarantee` *is* a structural primitive, it must *behave* as `flatMap` composed with `attempt`, no? I don't see how anything otherwise could be the case. What is the difference between a law and a test in this case?